### DISCUSSION TITLE
Is it possible to prefetch/process once for all pages when generating SSG pages? (during initial build)

### DISCUSSION THREAD

**Discussion by Vadorequest** [vercel/next.js#18000] [Help]
I wonder if there is a way to fetch only once and process some data and then make it available to all SSG pages, during the initial build.

Currently, every page fetches the same data and then performs its own data processing. Having 15 pages, this is becoming a pain point and causes issues with the API that's being hit way too much in a short amount of time (burst).

I'm looking for a way to optimize this greatly. I've performed some optimizations to avoid bursting the API by delaying the calls a bit, but a native solution is needed at this point.

I'm using Vercel to build the app.

**Comment by mrclay** [comment]
I’m considering a customized proxy cache in front of the API—or running in node during build.

**Comment by mrclay** [comment]
More crude: detect build-time generation and throw pauses before the API calls.

**Reply by Vadorequest** [reply]
That's what I already did. It works. It gets complicated in SSG "preview" mode though, because while pausing during the initial build is fine (a few seconds aren't gonna change much) it's not a great UX when people are actually waiting for the page to display. Preview mode and SSR are basically the same, from an end-user viewpoint.

Also, you still fetch once per page. I got 8 pages, 2 langs, which yields 16 statically generated pages, so 16 times where all API endpoints get fetched.

I did build a Vercel disk cache to reduce the amount of requests, and it works too. But it's still a bit bursting. Could be done better.

**Comment by fabb** [comment]
Have you tried `val-loader`?
https://github.com/vercel/next.js/discussions/10949#discussioncomment-90276

**Reply by Vadorequest** [reply]
@fabb I haven't yet, it definitely looks like I should use it to load the files that'll be generated.

But now, how/where do you recommend to generate those files? https://github.com/vercel/next.js/discussions/10949#discussioncomment-110385 seems to do it from the `next.config.js`, but I'm concerned about ES5+ support 😕 

I just need to call my function that will call the Airtable API and generate those JSON files, and then load those JSON files using `val-loader`. I'm just not sure where that API call should happen.
